nym-vpn-core/crates/nym-statistics/src/controller.rs

```diff
@@ -0,0 +1,198 @@
+    async fn disabled_loop(mut self) {
+        loop {
+            tokio::select! {
+                biased;
+                _ = self.cancel_token.cancelled() => {
+                    tracing::trace!("StatisticsController : Received cancellation signal");
+                    return;
+                },
+                stats_event = self.stats_rx.recv() => match stats_event {
+                        Some(_) => {},
+                        None => {
+                            tracing::trace!("StatisticsController: shutting down due to closed stats channel");
+                            return;
+                        }
+                },
+            }
+        }
+    }
+    async fn enabled_loop(mut self) {
+        if !self.config.enabled || self.stats_api_client.is_none() {
+            tracing::error!(
+                "StatisticsController : Enabled loop with disabled collection or missing api client. This should never happen."
+            );
+            return;
+        }
+
+        // Safety : We just checked that self.stats_api_client wasn't None
+        #[allow(clippy::unwrap_used)]
+        let stats_api_client = self.stats_api_client.unwrap();
+
+        let mut send_timer: Option<Pin<Box<Sleep>>> = None;
+
+        loop {
+            tokio::select! {
+                biased;
+                _ = self.cancel_token.cancelled() => {
+                    tracing::trace!("StatisticsController : Received cancellation signal");
+                    break;
+                },
+                stats_event = self.stats_rx.recv() => match stats_event {
+                    Some(stats_event) => self.handler.handle_event(stats_event),
+                    None => {
+                        tracing::trace!("StatisticsController: shutting down due to closed stats channel");
+                        break;
+                    }
+                },
+                Ok(_) = self.tunnel_state.changed() => {
+                    match self.tunnel_state.borrow().clone() {
+                        TunnelState::Connected {..} => {
+                            if let Err(e) = self.stats_tx.send(StatisticsEvent::new_connected()){
+                                tracing::warn!("Failed to send stat event : {e}")
+                            }
+                            let random_delay_secs = Uniform::new_inclusive(0, self.config.max_reporting_delay).sample(&mut rand::thread_rng());
+                            tracing::debug!("StatisticsController : Trying to send report in {random_delay_secs} secs");
+                            send_timer = Some(Box::pin(tokio::time::sleep(Duration::from_secs(random_delay_secs))));
+
+                        },
+                        TunnelState::Disconnecting {..}=>{
+                            if let Err(e) = self.stats_tx.send(StatisticsEvent::new_disconnecting()){
```

Comment:
FYI you can look into `after_disconnect` parameter if you want to get more insight.

```rs
pub enum ActionAfterDisconnect {
    Nothing,
    Reconnect,
    Offline,
    Error,
}
```